SLO Tags [GA = w/e June 9] (#2275)
SumoLogic/sumologic-documentation: docs/observability/reliability-management-slo/create-slo.md

  </tr>
</table>

   <img src={useBaseUrl('img/observability/slo-logs.gif')} alt="Reliability Management SLO SLI" />

8. Define your SLO for target amount and duration period to monitor:
   * **Target**. The value in percentage you want to target for the SLO (for example, you'd enter `99` for 99%).
   * **Compliance Type**. **Rolling** provides a sequence of recent days for the Compliance Period, such as the last 7d or last 30d. **Calendar** calculates over a window of time for a Week.
   * **Timezone**. Selecting a timezone is important to accurately assign events on the boundary of a compliance period, such as events received at 11:59 PM in a particular time zone.<br/><img src={useBaseUrl('img/observability/slo-create-slo.png')} alt="Reliability Management SLO SLI" width="400"/>
9. Enter SLO Details:
   * **Name**. Name of the SLO.
   * **Description**. Short explanation of your SLO.

>    * **Tags**. Tags are key/value pairs which associate SLO with a particular category. [Learn more here](#slo-tags-and-filters). These tags have additional metadata to describe an SLO beyond the name and description, which can help you categorize, search, and filter your SLOs.<br/>Select the key and its possible values from the tags dropdown, or create your own key/value pair for the tag.<br/>

Reviewer: how about

`
Tags are key/value pairs using which you can attach additional metadata to your SLOs beyond the name and description [Learn more here](#slo-tags-and-filters). Tags can help you categorize, search, and filter your SLOs.
`

>    * **Tags**. Tags are key/value pairs which associate SLO with a particular category. [Learn more here](#slo-tags-and-filters). These tags have additional metadata to describe an SLO beyond the name and description, which can help you categorize, search, and filter your SLOs.<br/>Select the key and its possible values from the tags dropdown, or create your own key/value pair for the tag.<br/>

Author: ```suggestion
   * **Tags**. Tags are key/value pairs that allow you to apply additional metadata to your SLOs beyond the name and description. [Learn more here](#slo-tags-and-filters). Tags can help you categorize, search, and filter your SLOs.<br/>Select the key and its possible values from the tags dropdown, or create your own key/value pair for the tag.<br/>
```

  :::info
  You can associate multiple tags with your SLO.
  :::
  <img src={useBaseUrl('img/observability/slo-details.png')} alt="SLO Details" width="800"/>
10. Click **Save**. To create a monitor, click [Save and Create Monitor](#create-an-slo-monitor).

### SLO Tags and Filters

> You can add key/value pair tags to your SLOs to allow you to better organize and filter them. You might find it useful to add tags for `team`, `service`, and `application`, for example.

Author: ```suggestion
You can add key/value pair tags to your SLOs to allow you to better organize and filter them. You might find it useful to add tags for `team`, `service`, and `application`, for example.

#### Limitations 

- Tag keys cannot start with the prefixes `sumo.` or `_`
- Tag keys can only contain letters, numbers, `_`, `.`, `/`, `+`, `-`, `@`
- Tag values can only contain letters, white space, numbers, `_`, `.`, `/`, `=`, `+`, `-`, `@`

You can associate a maximum of 50 tags per SLO.
```

> You can add key/value pair tags to your SLOs to allow you to better organize, sort, and filter them. As an example, you might find it useful to add variables such as environment, application name, and region.

Author: ```suggestion
You can add key/value pair tags to your SLOs to allow you to better organize and filter them. You might find it useful to add tags for `team`, `service`, and `application`, for example.
```

> You can add key/value pair tags to your SLOs to allow you to better organize, sort, and filter them. As an example, you might find it useful to add variables such as environment, application name, and region.

Reviewer: nit: similar comments as before i.e.
1. can remove sort
2. better examples for tags would be `team` `service` `application`

> #### Add a Tag

Reviewer: can we mention the following constraints

`Tag keys cannot start with the prefixes 'sumo.' or '_'`
`Tag keys can only contain letters, numbers, _, ., /, +, -, @`
`Tag values can only contain letters, white space, numbers, _, ., /, =, +, -, @`

`You can associate a maximum of 50 tags per SLO.`

To add a tag(s) to an existing SLO:

1. In Sumo Logic, click **Manage Data** > **Monitoring** > **SLOs** tab.
1. Click on any SLO line item in your list, then click **Edit**.
1. Scroll down to section **(3) SLO Details**. Click on **Tags (Optional)** and type in a new tag or select an existing tag.
1. Click **Save**.

To add a tag while creating a new SLO:

1. In Sumo Logic, click **Manage Data** > **Monitoring** > **SLOs** tab.
1. Click **Add** > **New SLO**.
1. After you've filled out sections **1** and **2**, scroll down to section **(3) SLO Details**. Click on **Tags (Optional)** and type in a new tag or select an existing tag.
1. Click **Save**.

#### Filter Tags

After you've added a tag, you'll see it populate in the **Tags** column next to your SLO in the list.

> To filter your SLOs by tag:

Reviewer: can we mention here that if you select multiple values for same tag key, they are `OR`ed and tag filters for different tag keys are `AND`ed ? Example

`service=ingest` `service=query` `team=metrics` - this tag filter translates to `service` is either `ingest` OR `query`, AND `team` is `metrics`

> To filter your SLOs by tag:

Reviewer: nit: `To filter your SLOs by tags` (tags is plural)

1. Click **Add a filter** at the top of the screen, then click **Tag**.<br/><img src={useBaseUrl('img/observability/slo-tags.png')} alt="slo-tags.png" />
1. Scroll through the list of tags or type in the tag name you're looking for.<br/><img src={useBaseUrl('img/observability/slo-tags.gif')} alt="slo-tags.gif" />

#### Use cases

> You can leverage SLO tags in **Log Search** queries. In this example, we run a query to display all SLOs. You'll see a **Tags** column here.

Reviewer: did you intend to include an SLO lookup query here i.e. `cat sumo;//content/slos` and an associated image?

Author: good catch!

  </tr>
</table>

   <img src={useBaseUrl('img/observability/slo-logs.gif')} alt="Reliability Management SLO SLI" />

8. Define your SLO for target amount and duration period to monitor:
   * **Target**. The value in percentage you want to target for the SLO (for example, you'd enter `99` for 99%).
   * **Compliance Type**. **Rolling** provides a sequence of recent days for the Compliance Period, such as the last 7d or last 30d. **Calendar** calculates over a window of time for a Week.
   * **Timezone**. Selecting a timezone is important to accurately assign events on the boundary of a compliance period, such as events received at 11:59 PM in a particular time zone.<br/><img src={useBaseUrl('img/observability/slo-create-slo.png')} alt="Reliability Management SLO SLI" width="400"/>
9. Enter SLO Details:
   * **Name**. Name of the SLO.
   * **Description**. Short explanation of your SLO.

>    * **Tags**. Tags are key/value pairs that allow you to apply additional metadata to your SLOs beyond the name and description. [Learn more here](#slo-tags-and-filters). Tags can help you categorize, search, and filter your SLOs.<br/>Select the key and its possible values from the tags dropdown, or create your own key/value pair for the tag.<br/>

Reviewer 2: `Tags can help you categorize, search, and filter your SLOs. They also enable you to correlate your other content items/data with SLOs.`

### SLO Tags and Filters

You can add key/value pair tags to your SLOs to allow you to better organize and filter them. You might find it useful to add tags for `team`, `service`, and `application`, for example.

#### Limitations

- Tag keys cannot start with the prefixes `sumo.` or `_`
- Tag keys can only contain letters, numbers, `_`, `.`, `/`, `+`, `-`, `@`
- Tag values can only contain letters, white space, numbers, `_`, `.`, `/`, `=`, `+`, `-`, `@`

You can associate a maximum of 50 tags per SLO.

#### Add a Tag

To add a tag(s) to an existing SLO:

1. In Sumo Logic, click **Manage Data** > **Monitoring** > **SLOs** tab.
1. Click on any SLO line item in your list, then click **Edit**.
1. Scroll down to section **(3) SLO Details**. Click on **Tags (Optional)** and type in a new tag or select an existing tag.
1. Click **Save**.

To add a tag while creating a new SLO:

1. In Sumo Logic, click **Manage Data** > **Monitoring** > **SLOs** tab.
1. Click **Add** > **New SLO**.
1. After you've filled out sections **1** and **2**, scroll down to section **(3) SLO Details**. Click on **Tags (Optional)** and type in a new tag or select an existing tag.
1. Click **Save**.

> #### Filter SLOs By Tags

Reviewer 2: We should add a section for "Tags in SLO lookup table". Tags are available under SLO lookup table. Users will be able to see tags for SLOs under lookup table content. Screenshot (query [link](https://epddata.stag.sumologic.net/ui/#/search/create?id=cmcX2e0hjZXOl7EoQSBYslMHUi2dJd4XUAR0fjmW)):

<img width="1454" alt="image" src="https://github.com/SumoLogic/sumologic-documentation/assets/13195152/8cb5e730-54ae-477a-912f-576a396e9924">
 
To see all the SLOs with at least one tag ([link](https://epddata.stag.sumologic.net/ui/#/search/create?id=bKESghyFSBsxeeEkQHNhH4qTKE0j2N4ZD6j1OHhe)):

<img width="1437" alt="image" src="https://github.com/SumoLogic/sumologic-documentation/assets/13195152/cfcf66d3-1848-4a8a-b879-c7a0015be8a0">

Users can use tags in this lookup table for correlation with other data in sumo logic. e.g. they can use it for finding SLO output data for all SLOs that belong to service `ingestion` using the following query:

_view=sumologic_slo_output | lookup tags from sumo://content/slos on sloId=sloId | json field=tags "service" | where service="ingestion"

1. Click **Save**.

To add a tag while creating a new SLO:

1. In Sumo Logic, click **Manage Data** > **Monitoring** > **SLOs** tab.
1. Click **Add** > **New SLO**.
1. After you've filled out sections **1** and **2**, scroll down to section **(3) SLO Details**. Click on **Tags (Optional)** and type in a new tag or select an existing tag.
1. Click **Save**.

#### Filter Tags

After you've added a tag, you'll see it populate in the **Tags** column next to your SLO in the list.

1. Click **Add a filter** at the top of the screen, then click **Tag**.<br/><img src={useBaseUrl('img/observability/slo-tags.png')} alt="slo-tags.png" width="400"/>
1. Scroll through the list of tags or type in the tag name you're looking for.<br/><img src={useBaseUrl('img/observability/slo-tags.gif')} alt="slo-tags.gif" />

If you run a query with multiple values for same tag key, they are `OR`'d. Tag filters for different tag keys are `AND`'d.

In this tag filter example query below, it's looking for SLOs where the `app` is either `sumo+1` OR `sumologic`, AND the `event` is `api-call`.

<img src={useBaseUrl('img/observability/slo-tags-query.png')} alt="slo-tags-query.png" width="500"/>

> #### Use cases

Author: Good thinking. Just fixed. Can you take a look?

<img src={useBaseUrl('img/observability/slo-tags-query.png')} alt="slo-tags-query.png" width="500"/>

### SLO as Log Messages

Sumo Logic continuously computes data for your SLO behind the scenes. This data, which powers your SLO dashboard, is also made available as log messages that conform to the following schema:

* `Time`: timestamp
* `sloId`: Id of the SLO, as displayed in the SLO dashboard URL
* `goodCount`: count of good requests, for request-based, and good windows for windows-based SLOs, based on SLO query definition
* `totalCount`: count of eligible requests for request-based, and eligible windows for windows-based SLOs, based SLO query definition
* `sloVersion`: version of SLO definition

View the schema by executing the following query:

```sql
_view=sumologic_slo_output sloId="<your-SLO-ID>"
| where [subquery: _view=sumologic_slo_output sloId="<your-SLO-ID>"
| max(sloVersion) as sloVersion | compose sloVersion]
-- (replace with a valid SLO Id)
```

These log messages will be delayed by one hour, as the system ensures consistency to account for ingest delay of source telemetry.

### SLO Lookup Tables

> You can call a SLO Lookup Table to view all SLO metadata in your environment. These tables reside under a fixed path, `sumo://content/slos`. Data is managed and refreshed automatically on our end.

Reviewer 2: nit: we can replace `call` by `query`.

> You can call a SLO Lookup Table to view all SLO metadata in your environment. These tables reside under a fixed path, `sumo://content/slos`. Data is managed and refreshed automatically on our end.

Author: ```suggestion
You can query a SLO Lookup Table to view all SLO metadata in your environment. These tables reside under a fixed path, `sumo://content/slos`. Data is managed and refreshed automatically on our end.
```

1. In the search field, enter the following snippet. This will join data from multiple sources for your lookup table.
  ```sql
  _view=sumologic_slo_output
  | lookup * from sumo://content/slos on sloId, sloVersion
  | where !isBlank (sloname) and slofolderpath matches "*"
  | concat (sloname, " (", sloId, ")") as sloUniqueName
  | sum (goodCount) as goodEvents, sum(totalCount) as totalEvents, last (compliancetarget) as target, last(slofolderpath) as sloPath, last(sliwindowsize) as sliwindowsize, last(slievaluationtype) as evaluationType by sloUniqueName
  | totalEvents - goodEvents as badEvents
  | if (evaluationType = "Window", queryTimeRange() / 1000 / sliwindowsize, totalEvents) as denominator
  | 100 * (1 - badEvents / denominator) as sli
  | 100 * (sli - target) / (100 - target) as budgetRemaining
  | fields sloUniqueName, budgetRemaining
  ```
1. Click **Add to Dashboard**.<br/><img src={useBaseUrl('img/observability/add-to-dashboard.png')} alt="add-to-dashboard" width="200"/>

#### Example 2

You can leverage SLO tags in **Log Search** queries. Here's a query, for example, to display all SLOs that have tags:

> <img src={useBaseUrl('img/observability/slo-tags-query-log.png')} alt="slo-tags-query.png" /><br/>

Author: Hi @tarunk2 , about [this comment](https://github.com/SumoLogic/sumologic-documentation/pull/2275/files#r1214309385), I did have include a bit about tags in SLO lookup tables further down here in the doc. I'll combine the example you shared with mine and then cross-link those sections.
